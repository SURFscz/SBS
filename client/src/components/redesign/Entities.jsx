import React from "react";
import "./Organisations.scss";
import I18n from "../../locale/I18n";
import {ReactComponent as SearchIcon} from "@surfnet/sds/icons/functional-icons/search.svg";
import Button from "../Button";
import PropTypes from "prop-types";
import {isEmpty, sortObjects, valueForSort} from "../../utils/Utils";
import {headerIcon} from "../../forms/helpers";
import "./Entities.scss";
import SpinnerField from "./SpinnerField";
import Explain from "../Explain";
import {Pagination} from "@surfnet/sds";
import {pageCount} from "../../utils/Pagination";

class Entities extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            query: "",
            sorted: props.defaultSort,
            reverse: false,
            showExplanation: false,
            page: 1
        }
    }

    componentDidMount = () => {
        setTimeout(() => this.props.inputFocus && this.input && this.input.focus(), 150);
    }

    newEntity = () => {
        const {newEntityPath, newEntityFunc} = this.props;
        if (newEntityFunc) {
            newEntityFunc();
        } else {
            this.props.history.push(newEntityPath);
        }
    };

    closeExplanation = () => this.setState({showExplanation: false});

    queryChanged = e => {
        const query = e.target.value;
        this.setState({query: query});
        const {searchCallback, customSearch, entities, searchAttributes} = this.props;
        if (customSearch) {
            customSearch(query);
        }
        if (searchCallback) {
            searchCallback(this.filterEntities(entities, query, searchAttributes));
        }
    }

    renderSearch = (modelName, title, entities, query, searchAttributes, showNew, newLabel, filters, explain, customSearch, hideTitle) => {
        const filterClassName = !hideTitle && filters ? "filters-with-title" : `${modelName}-search-filters`;
        return (
            <section className="entities-search">
                {showNew &&
                <Button onClick={this.newEntity}
                        className={`${hideTitle && !filters ? "no-title" : ""}`}
                        txt={newLabel || I18n.t(`models.${modelName}.new`)}/>
                }
                {!hideTitle && <h2>{title || `${I18n.t(`models.${modelName}.title`)} (${entities.length})`}</h2>}
                <div className={filterClassName}>{filters}</div>
                <div className={`search ${showNew ? "" : "standalone"}`}>
                    {(!isEmpty(searchAttributes) || customSearch) &&
                    <div className={"sds--text-field sds--text-field--has-icon"}>
                        <div className="sds--text-field--shape">
                            <div className="sds--text-field--input-and-icon">
                                <input className={"sds--text-field--input"}
                                       type="search"
                                       onChange={this.queryChanged}
                                       ref={ref => this.input = ref}
                                       value={query}
                                       placeholder={I18n.t(`models.${modelName}.searchPlaceHolder`)}/>
                                <span className="sds--text-field--icon">
                                    <SearchIcon/>
                                </span>
                            </div>
                        </div>
                    </div>}

                </div>
            </section>
        );
    };

    filterEntities = (entities, query, searchAttributes, customSearch) => {
        if (isEmpty(query) || customSearch) {
            return entities;
        }
        query = query.toLowerCase();
        return entities.filter(entity => {
            return searchAttributes.some(attr => {
                const val = valueForSort(attr, entity);
                return val.toLowerCase().indexOf(query) > -1
            });
        });

    };

    setSorted = key => () => {
        const {sorted, reverse} = this.state;
        const reversed = (sorted === key ? !reverse : false);
        this.setState({sorted: key, reverse: reversed})
    }

    getEntityValue = (entity, column) => {
        if (column.mapper) {
            return column.mapper(entity);
        }
        return entity[column.key];
    }

    onRowClick = (rowLinkMapper, entity) => e => {
        const hasValue = typeof rowLinkMapper === "function" && rowLinkMapper(entity);
        if (hasValue) {
            hasValue(entity)(e);
        }
    }

    renderEntities = (entities, sorted, reverse, modelName, tableClassName, columns, children,
                      rowLinkMapper, customNoEntities, onHover, actions, showActionsAlways, actionHeader, page, pagination) => {
        const hasEntities = !isEmpty(entities);
        const total = entities.length;
        if (pagination) {
            const minimalPage = Math.min(page, Math.ceil(entities.length / pageCount));
            entities = entities.slice((minimalPage - 1) * pageCount, minimalPage * pageCount);
        }
        return (
            <section className="entities-list">
                {(actions && (showActionsAlways || hasEntities)) && <div className={`actions-header ${actionHeader}`}>
                    {actions}
                </div>}
                {hasEntities &&
                <div className={"sds--table"}>
                    <table className={tableClassName || modelName}>
                        <thead>
                        <tr>
                            {columns.map((column, i) => {
                                const showHeader = !actions || i < 2 || column.showHeader;
                                return <th key={`th_${column.key}_${i}`}
                                           className={`${column.key} ${column.class || ""} ${column.nonSortable ? "" : "sortable"} ${showHeader ? "" : "hide"}`}
                                           onClick={this.setSorted(column.key)}>
                                    {column.header}
                                    {headerIcon(column, sorted, reverse)}
                                </th>
                            })}
                        </tr>
                        </thead>
                        <tbody>
                        {entities.map((entity, index) =>
                            <tr key={`tr_${entity.id}_${index}`}
                                className={`${(typeof rowLinkMapper === "function" && rowLinkMapper(entity)) ? "clickable" : ""} ${onHover ? "hoverable" : ""}`}>
                                {columns.map((column, i) =>
                                    <td key={`td_${column.key}_${i}`}
                                        onClick={(column.key !== "check" && column.key !== "role" && !column.hasLink) ?
                                            this.onRowClick(rowLinkMapper, entity) : undefined}
                                        className={`${column.key} ${column.nonSortable ? "" : "sortable"} ${column.className ? column.className : ""}`}>
                                        {this.getEntityValue(entity, column)}
                                    </td>)}
                            </tr>
                        )}
                        </tbody>
                    </table>
                </div>}
                {(!hasEntities && !children) &&
                <p className="no-entities">{customNoEntities || I18n.t(`models.${modelName}.noEntities`)}</p>}
                {pagination && <Pagination currentPage={page}
                                           onChange={nbr => this.setState({page: nbr})}
                                           total={total}
                                           pageCount={pageCount}/>}
            </section>
        );
    };

    render() {
        const {
            modelName,
            entities,
            showNew,
            newLabel,
            searchAttributes,
            columns,
            children,
            loading,
            customSearch,
            actions,
            title,
            filters,
            explain,
            rowLinkMapper,
            tableClassName,
            explainTitle,
            className = "",
            customNoEntities,
            hideTitle,
            onHover,
            actionHeader = "",
            pagination = true,
            showActionsAlways,
            displaySearch = true
        } = this.props;
        if (loading) {
            return <SpinnerField/>;
        }
        const {query, sorted, reverse, page, showExplanation} = this.state;
        const filteredEntities = this.filterEntities(entities, query, searchAttributes, customSearch);
        const sortedEntities = sortObjects(filteredEntities, sorted, reverse);
        return (
            <div className={`mod-entities ${className}`}>
                {explain && <Explain
                    close={this.closeExplanation}
                    subject={explainTitle || I18n.t("explain.services")}
                    isVisible={showExplanation}>
                    {explain}
                </Explain>}
                {displaySearch && this.renderSearch(modelName, title, entities, query, searchAttributes, showNew, newLabel, filters, explain, customSearch, hideTitle)}
                {this.renderEntities(sortedEntities, sorted, reverse, modelName, tableClassName, columns, children,
                    rowLinkMapper, customNoEntities, onHover, actions, showActionsAlways, actionHeader, page, pagination)}
                <div>{this.props.children}</div>
            </div>);
    }
}

Entities.propTypes = {
    title: PropTypes.string,
    entities: PropTypes.array.isRequired,
    modelName: PropTypes.string.isRequired,
    tableClassName: PropTypes.string,
    className: PropTypes.string,
    customNoEntities: PropTypes.string,
    searchAttributes: PropTypes.array,
    defaultSort: PropTypes.string.isRequired,
    loading: PropTypes.bool.isRequired,
    columns: PropTypes.array.isRequired,
    newEntityPath: PropTypes.string,
    newEntityFunc: PropTypes.func,
    onHover: PropTypes.bool,
    displaySearch: PropTypes.bool,
    rowLinkMapper: PropTypes.func,
    searchCallback: PropTypes.func,
    customSearch: PropTypes.func,
    showNew: PropTypes.bool,
    actions: PropTypes.any,
    showActionsAlways: PropTypes.bool,
    filters: PropTypes.any,
    explain: PropTypes.any,
    inputFocus: PropTypes.bool,
    hideTitle: PropTypes.bool
};

export default Entities;
