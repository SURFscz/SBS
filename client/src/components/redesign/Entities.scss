@import "../../stylesheets/vars";

.mod-entities {

    max-width: $medium;
    width: 100%;
    margin: 0 auto;

    @media (max-width: $tablet-max) {
        padding: 0 15px;
    }


    &:not(.first) {
        margin-bottom: 25px;
    }

    .entities-search {
        display: flex;
        align-items: center;
        padding: 32px 0 28px 0;
        min-height: 40px;
        justify-content: space-between;

        @media (max-width: $compact-medium) {
            flex-direction: column;
            align-items: normal;
        }

        h2 {
            @media (max-width: $compact-medium) {
                margin-bottom: 15px
            }

        }

        div.filter-select__value-container {
            min-height: 46px;
        }

        div.filter-select__control {
            min-height: 46px;
            border-color: var(--sds--color--gray--400);
        }

        div.filters-with-title {
            margin-left: auto;
        }

        svg.help {
            color: $blue;
            font-size: 14px;
            cursor: pointer;
            margin: 0 8px 18px auto;

            &:hover {
                color: darken($blue, 10%);
            }
        }

        .search {
            position: relative;
            display: flex;

            &.standalone {
                margin-left: auto;
            }

            @media (max-width: $compact-medium) {
                margin-left: 0;
            }

            input[type=search] {
                flex-grow: 2;
                min-width: 360px;

                @media (max-width: $compact-medium) {
                    min-width: 0;
                }
            }

            .sds--text-field--icon svg {
                width: calc(15em / 16);
                height: calc(15em / 16)
            }

        }

    }

    section.entities-list {
        position: relative;

        div.actions-header {
            position: absolute;
            left: 76px;
            top: 2px;

            &.collaboration-services {
                left: 32px;
            }

            &.collaboration-groups {
                width: 100%;
                left: 0;
                top: -2px;
                position: relative;
            }

            &.adjust-top {
                top: 65px;
            }
        }

        .sds--pagination {
            margin-top: 25px;
        }
    }

    svg.impersonate {
        width: 40px;
        height: auto;
        cursor: pointer;
        fill: #737373;
        padding: 5px;
    }


    table {
        width: 100%;
        table-layout: fixed;

        thead {
            th {

                &.sortable {
                    cursor: pointer;
                }

                &.icon {
                    width: 64px;
                    padding-right: 0;
                    padding-left: 0;
                }

                &.logo {
                    width: 85px;
                }

                svg.fa-caret-up, svg.fa-caret-down {
                    margin-left: 5px;
                }

                &.checker {
                    width: 35px;
                }

                &.center {
                    text-align: center;
                }

                &.right {
                    text-align: right;
                }

                &.hide {
                    color: white;
                    cursor: default;
                }

                @media (max-width: $compact-medium) {
                    &.status, &.requester, &.impersonate, &.logo, &.collaboration_memberships_count,
                    &.invitations_count, &.organisation_memberships_count, &.collaborations_count,
                    &.connection_requests_count, &.icon, &.user__schac_home_organisation {
                        display: none;
                    }
                }

            }


        }

        tbody {
            tr {
                &.clickable {
                    cursor: pointer;
                }

                .action-icons-container {
                    display: flex;
                    align-items: center;
                    justify-content: flex-end;
                }

                .admin-icons {
                    display: none;
                    margin: auto 0;

                    svg {
                        color: $dark-grey;
                        cursor: pointer;
                        margin: 0 5px;
                        width: 20px;
                        height: auto;

                        &:hover {
                            color: black;
                        }
                    }
                }

                &.hoverable {
                    &:hover {
                        div.admin-icons {
                            display: flex;
                        }
                    }
                }
            }

            td {
                word-break: break-word;
                vertical-align: middle;

                &.center {
                    text-align: center;
                }

                &.right {
                    text-align: right;
                }

                &.icon {
                    text-align: center;
                    padding-left: 0;
                    padding-right: 0;

                    div.member-icon {
                        display: flex;
                        justify-content: space-around;

                        svg {
                            width: 24px;
                        }
                    }
                }

                &.logo {
                    padding: 8px 0 4px 0;

                    img {
                        border-radius: 4px;
                        //min-width: 80px;
                        //min-height: 58px;
                        //max-width: 80px;
                        //max-height: 58px;
                        width: 80px;
                        height: auto;
                        margin-right: 5px;
                    }
                }

                @media (max-width: $compact-medium) {
                    &.status, &.requester, &.impersonate, &.logo, &.collaboration_memberships_count,
                    &.invitations_count, &.organisation_memberships_count, &.connection_requests_count,
                    &.collaborations_count, &.icon, &.user__schac_home_organisation {
                        display: none;
                    }
                }

                .sds--chips span {
                    white-space: nowrap;
                }
            }
        }
    }

    div.user-name-email {
        display: flex;
        flex-direction: column;

        span.email {
            color: $dark-grey;
        }
    }

    div.chip-container {

        .sds--chips {
            margin: 0 auto;
        }
    }

    .sds--pagination a {
        font-weight: normal;
    }

}


