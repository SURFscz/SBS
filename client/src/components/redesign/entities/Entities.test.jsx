import React from "react";
import {render} from "@testing-library/react";
import Entities from "./Entities";

const columns = [
    {name: "name", label: "Name"}
];

describe("Entities", () => {
    it("matches snapshot with default props", () => {
        const {asFragment} = render(
            <Entities
                entities={[]}
                onEntityClick={() => {
                }}
                onCreateEntity={() => {
                }}
                user={{admin: true}}
                columns={columns}
                modelName="modelName"
                defaultSort="asc"
                loading={false}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
