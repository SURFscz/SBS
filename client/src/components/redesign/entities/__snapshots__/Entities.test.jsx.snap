// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entities matches snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="mod-entities "
  >
    <section
      class="entities-search"
    >
      <h2>
        [missing "models.modelName.title" translation] (0)
      </h2>
      <div
        class="search-filters modelName-search-filters"
      />
      <div
        class="search standalone"
      />
    </section>
    <section
      class="entities-list"
    />
    <div />
  </div>
</DocumentFragment>
`;
